sql/core/src/test/scala/org/apache/spark/sql/connector/WriteDistributionAndOrderingSuite.scala

```diff
@@ -0,0 +1,619 @@
+      case "overwrite" => execute(writer.overwrite(lit(true)))
+      case "overwriteDynamic" => execute(writer.overwritePartitions())
+    }
+
+    checkPartitioningAndOrdering(executedPlan, expectedWritePartitioning, expectedWriteOrdering)
+
+    checkAnswer(spark.table(tableNameAsString), df)
+  }
+
+  private def checkPartitioningAndOrdering(
+      plan: SparkPlan,
+      partitioning: physical.Partitioning,
+      ordering: Seq[catalyst.expressions.SortOrder]): Unit = {
+
+    val sorts = plan.collect { case s: SortExec => s }
+    assert(sorts.size <= 1, "must be at most one sort")
+    val shuffles = plan.collect { case s: ShuffleExchangeLike => s }
+    assert(shuffles.size <= 1, "must be at most one shuffle")
+
+    val actualPartitioning = plan.outputPartitioning
+    val expectedPartitioning = partitioning match {
+      case p: physical.RangePartitioning =>
+        val resolvedOrdering = p.ordering.map(resolveAttrs(_, plan))
+        p.copy(ordering = resolvedOrdering.asInstanceOf[Seq[catalyst.expressions.SortOrder]])
+      case p: physical.HashPartitioning =>
+        val resolvedExprs = p.expressions.map(resolveAttrs(_, plan))
+        p.copy(expressions = resolvedExprs)
+      case other => other
+    }
+    assert(actualPartitioning == expectedPartitioning, "partitioning must match")
+
+    val actualOrdering = plan.outputOrdering
+    val expectedOrdering = ordering.map(resolveAttrs(_, plan))
+    assert(actualOrdering == expectedOrdering, "ordering must match")
+  }
+
+  private def resolveAttrs(
+      expr: catalyst.expressions.Expression,
+      plan: SparkPlan): catalyst.expressions.Expression = {
+
+    expr.transform {
+      case UnresolvedAttribute(Seq(attrName)) =>
+        plan.output.find(attr => resolver(attr.name, attrName)).get
+      case UnresolvedAttribute(nameParts) =>
+        val attrName = nameParts.mkString(".")
+        fail(s"cannot resolve a nested attr: $attrName")
+    }
+  }
+
+  private def attr(name: String): UnresolvedAttribute = {
+    UnresolvedAttribute(name)
+  }
+
+  private def catalog: InMemoryTableCatalog = {
+    val catalog = spark.sessionState.catalogManager.catalog("testcat")
+    catalog.asTableCatalog.asInstanceOf[InMemoryTableCatalog]
+  }
+
+  // executes a write operation and keeps the executed physical plan
+  private def execute(writeFunc: => Unit): SparkPlan = {
+    var executedPlan: SparkPlan = null
+
+    val listener = new QueryExecutionListener {
+      override def onSuccess(funcName: String, qe: QueryExecution, durationNs: Long): Unit = {
+        executedPlan = qe.executedPlan
+      }
+      override def onFailure(funcName: String, qe: QueryExecution, exception: Exception): Unit = {
+      }
+    }
+    spark.listenerManager.register(listener)
+
+    writeFunc
+
+    sparkContext.listenerBus.waitUntilEmpty()
+
+    executedPlan match {
+      case w: V2TableWriteExec =>
+        w.query match {
+          case p: AdaptiveSparkPlanExec => p.inputPlan
```

Comment:
I am using `inputPlan` here so that I can easily check for the number of shuffles by collecting `ShuffleExchangeLike` nodes. Otherwise, I'd have to deal with `ShuffleQueryStageExec` and other rewrites from AQE.

Reply:
FYI: there is an `AdaptiveSparkPlanHelper` to easily collect/traverse query plans with these AQE nodes stripped.

Reply:
Moved to `AdaptiveSparkPlanHelper`. Resolving this comment. Thanks, @cloud-fan!